utilities_test.go

```diff
@@ -126,3 +127,55 @@ func TestOuterHtml(t *testing.T) {
 		}
 	}
 }
+
+func TestRender(t *testing.T) {
+	doc, err := NewDocumentFromReader(strings.NewReader(allNodes))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	n0 := doc.Nodes[0]
+	nDT := n0.FirstChild
+	sMeta := doc.Find("meta")
+	sP := doc.Find("p")
+	nP := sP.Get(0)
+	nComment := nP.FirstChild
+	nText := nComment.NextSibling
+	sHeaders := doc.Find(".header")
+	sEmpty := doc.Find(".empty")
+
+	cases := []struct {
+		node *html.Node
+		sel  *Selection
+		want string
+	}{
+		{nDT, nil, "<!DOCTYPE html>"}, // render makes DOCTYPE all caps
+		{nil, sMeta, `<meta a="b"/>`}, // and auto-closes the meta
+		{nil, sP, `<p><!-- this is a comment -->
+		This is some text.
+		</p>`},
+		{nComment, nil, "<!-- this is a comment -->"},
+		{nText, nil, `
+		This is some text.
+		`},
+		{nil, sHeaders, `<h1 class="header"></h1>`},
+		{nil, sEmpty, ""}, // empty selection
+	}
```

Comment:
Would be nice to do a full document rendering, just to make sure (i.e. sel = doc.Selection).

Reply:
Added. However I had to slightly change `allNodes`, as `Render` seems to loose some of formatting and capitalize "doctype". I also made it const as there seems to be no reason to keep it variable.

Reply:
:+1: cool, yeah that's fine, the html rendering "canonicalizes" the representation and whitespace is not significant between the html tag and head/body so it is lost. Good call for the `const`.